libcnb/src/layer/public_interface.rs

```diff
@@ -0,0 +1,192 @@
+    ///
+    /// This can be used to invalidate a layer based on metadata. For example, the layer metadata
+    /// could contain a language runtime version string. If the version requested by the user is
+    /// different, this method should return true, causing the new language runtime version to be
+    /// installed.
+    ///
+    /// When not implemented, the layer will never be recreated.
+    ///
+    /// # Implementation Requirements
+    /// Implementations **MUST** be read-only. They **MUST NOT** modify the file-system or write
+    /// anything to stdout/stdout or any other stream.
+    fn should_be_recreated(
+        &self,
+        context: &BuildContext<Self::Buildpack>,
+        layer_data: &LayerData<Self::Metadata>,
+    ) -> Result<bool, <Self::Buildpack as Buildpack>::Error> {
+        Ok(false)
+    }
+
+    /// This method will be called by libcnb to determine if, based on the current state, the layer
+    /// needs to be updated. A layer will only be updated if it was restored from the cache and this
+    /// method returns `true`.
+    ///
+    /// If the layer was restored from cache and this method returns `false`, neither
+    /// [`create`](Self::create) nor [`update`](Self::update) will be called by libcnb. In this
+    /// case, the layer will stay unmodified, but libcnb will ensure its types match the ones
+    /// returned from the [`types`](Self::types) method.
+    ///
+    /// When not implemented, the layer will never be updated.
+    ///
+    /// # Implementation Requirements
+    /// Implementations **MUST** be read-only. They **MUST NOT** modify the file-system or write
+    /// anything to stdout/stdout or any other stream.
+    fn should_be_updated(
+        &self,
+        context: &BuildContext<Self::Buildpack>,
+        layer_data: &LayerData<Self::Metadata>,
+    ) -> Result<bool, <Self::Buildpack as Buildpack>::Error> {
+        Ok(false)
+    }
+
+    /// Updates the layer contents and metadata based on the cached version of a previous run.
+    ///
+    /// `layer_path` will be a directory with the data from a previous run. This method can modify
+    /// the contents freely. Layer metadata, including environment variables, is part of the return
+    /// value of this function and will be written to the appropriate locations by libcnb
+    /// automatically.
+    ///
+    /// The return value of this method is the canonical value for metadata and environment variables.
+    /// If the returned [`LayerResult`](LayerResult) does not contain metadata or environment
+    /// variables, the resulting layer will not have either. To keep the values from the cached
+    /// version you need to explicitly add them to the result. This can be done by reading that
+    /// data from the given [`LayerData`](LayerData) value.
+    ///
+    /// # Implementation Requirements
+    /// Implementations **MUST NOT** write to any other location than `layer_path`.
+    fn update(
```

Comment:
At the moment, if a buildpack implements `Layer` with `types()` that returns `cache=false`, they can still choose to implement `update()`, `should_be_updated()`, `should_be_recreated()` etc.

I wonder if we should create some more opinionated traits alongside `Layer` (such as `UncachedLayer` which only has `create()`) to allow this to be modelled more accurately?

Reply:
I think this `Layer` implementation should stay as-is in regards to the ability to provide an `update` implementation even if `types` always returns `cache = false` as it is the most accurate representation of reality.

Consider:
- The layer might have been written with `cache = true` by an older `Layer` implementation.
- The layer might have been written by a non-libcnb buildpack.
- The layer might have been modified by code outside of this `Layer` implementation. We can't guarantee that all layer modification happens inside `Layer`. It's probably an anti-pattern in a `libcnb`, but unpreventable.
- Since `layer` is a function, we can't enforce that it returns a constant value during buildpack execution and therefore cannot enforce `cache = false` unless we remove the ability to return a value for `cache`.
- If the layer is `handle_layer`'d multiple times, updating is a valid operation, even if the layer wasn't and will not be restored by the CNB lifecycle.

However, the above does not forbid more specialised implementations that enforce some of the restrictions you talk about:

```rust
/// Pseudo code:
trait UncachedLayer {
    type Buildpack: Buildpack;
    type Metadata: DeserializeOwned + Serialize + Clone;

    fn build() -> bool;
    fn launch() -> bool;

    fn create(
        &self,
        context: &BuildContext<Self::Buildpack>,
        layer_path: &Path,
    ) -> Result<LayerResult<Self::Metadata>, <Self::Buildpack as Buildpack>::Error>;
}

impl<A: UncachedLayer> Layer for A {
    type Buildpack = A::Buildpack;
    type Metadata = A::Metadata;

    fn types(&self) -> LayerTypes {
        LayerTypes {
            build: self.build(),
            launch: self.launch(),
            cache: false,
        }
    }

    fn create(
        &self,
        context: &BuildContext<Self::Buildpack>,
        layer_path: &Path,
    ) -> Result<LayerResult<Self::Metadata>, <Self::Buildpack as Buildpack>::Error> {
        self.create(context, layer_path)
    }
}
```

Note how we're able to implement this nicely on-top of the more powerful abstraction `Layer`. I suggest we move this discussion out of this PR and implement the more restrictive variants (if desired) separately.

Reply:
I agree that there are valid (although unlikely) cases where the layer could already exist (eg transitioning from cache=true to cache=false for a particular buildpack). However in those cases, I think libcnb should normalise these and purge the existing layer, like it normalises other cases.

Though I agree this is worth adding in more specialised implementations on top.

>     * Since `layer` is a function

I presume this should say `types`?

Reply:
> I presume this should say types?

Yes. Fixed. :)

Reply:
> I think libcnb should normalise these and purge the existing layer, like it normalises other cases.

I guess it's a matter of where we draw the line in terms of being opinionated that some use-cases are not supported. I don't think we have much precedent of intentionally not supporting a use-case from the spec apart from the "metadata without layer directory" case.

If we can reasonably support a use-case I think we should. I think that we should not remove support for certain use-case in `Layer` just because we think we shouldn't support them. There is no technical or API design need to do that. Especially at this point where none of us has written complex real-world buildpack with this framework yet, this feels like an arbitrary decision.

However, as said earlier, adding more opinionated traits on top of the more general one and encouraging their use via documentation, examples and tutorials is something I support.